sockeye/layers.py

```diff
@@ -58,18 +61,376 @@ def moments(inputs: mx.sym.Symbol) -> Tuple[mx.sym.Symbol, mx.sym.Symbol]:
+    def normalize(self, inputs: mx.sym.Symbol, eps: float = 0.000001) -> mx.sym.Symbol:
         """
-        Normalizes hidden units of inputs.
+        Normalizes hidden units of inputs as follows::
+
         inputs = scale * (inputs - mean) / sqrt(var + eps) + shift
 
         :param inputs: Inputs to normalize. Shape(batch_size, num_hidden).
         :param eps: Variance epsilon.
         :return: inputs_norm: Normalized inputs. Shape(batch_size, num_hidden).
         """
         mean, var = self.moments(inputs)
-        inputs_norm = mx.sym.broadcast_minus(inputs, mean, name='%s_inp_minus_mean' % self.prefix)
-        inputs_norm = mx.sym.broadcast_mul(inputs_norm, mx.sym.rsqrt(var + eps), name='%s_inp_norm' % self.prefix)
-        inputs_norm = mx.sym.broadcast_mul(inputs_norm, self.scale, name='%s_inp_norm_scaled' % self.prefix)
-        inputs_norm = mx.sym.broadcast_add(inputs_norm, self.shift, name='%s_inp_norm_scaled_shifted' % self.prefix)
+        inputs_norm = mx.sym.broadcast_minus(inputs, mean, name='%sinp_minus_mean' % self.prefix)
+        inputs_norm = mx.sym.broadcast_mul(inputs_norm, mx.sym.rsqrt(var + eps), name='%sinp_norm' % self.prefix)
+        inputs_norm = mx.sym.broadcast_mul(inputs_norm, self.scale, name='%sinp_norm_scaled' % self.prefix)
+        inputs_norm = mx.sym.broadcast_add(inputs_norm, self.shift, name='%sinp_norm_scaled_shifted' % self.prefix)
         return inputs_norm
+
+
+def split_heads(x: mx.sym.Symbol, length: int, heads: int) -> mx.sym.Symbol:
+    """
+    Returns a symbol with head dimension folded into batch and depth divided by the number of heads.
+
+    :param x: Symbol of shape (batch, length, depth).
+    :param length: Sequence length.
+    :param heads: Number of heads.
+    :return: Symbol of shape (batch * heads, length, depth_per_heads).
+    """
+    # (batch, length, heads, depth_per_head)
+    x = mx.sym.reshape(data=x, shape=(0, length, heads, -1))
+    # (batch, heads, length, depth/heads)
+    x = mx.sym.transpose(data=x, axes=(0, 2, 1, 3))
+    # (batch * heads, length, depth/heads)
+    return mx.sym.reshape(data=x, shape=(-3, length, -1))
+
+
+def combine_heads(x: mx.sym.Symbol, length: int, heads: int) -> mx.sym.Symbol:
+    """
+    Returns a symbol with both batch & length, and head & depth dimensions combined.
+
+    :param x: Symbol of shape (batch * heads, length, depth_per_head).
+    :param length: Sequence length.
+    :param heads: Number of heads.
+    :return: Symbol of shape (batch * length, depth).
+    """
+    # (batch, heads, length, depth_per_head)
+    x = mx.sym.reshape(data=x, shape=(-4, -1, heads, length, 0))
+    # (batch, length, heads, depth_per_head)
+    x = mx.sym.transpose(x, axes=(0, 2, 1, 3))
+    # (batch, length, depth)
+    return mx.sym.reshape(x, shape=(-1, length, -3))
+
+
+def broadcast_lengths(x: mx.sym.Symbol, heads: int) -> mx.sym.Symbol:
+    """
+    Broadcasts the length information of each sequence to multiple heads.
```

Comment:
I fail to see how this is related to "length". So you basically broadcast according to the number of heads. Where does length come in?